### PR TITLE
scheduler: write command should acquire write lock

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

## What have you changed? (mandatory)

Check lock type for write command in scheduler, to prevent adding a new write command but forget to acquire write lock. 

## What are the type of the changes? (mandatory)

- Improvement

## How has this PR been tested? (mandatory)

UT

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -207,6 +207,10 @@ impl<E: Engine> Scheduler<E> {
         let tctx = {
             let cmd = task.cmd();
             let lock = self.gen_lock(cmd);
+            // Write command should acquire write lock.
+            if !cmd.readonly() && !lock.is_write_lock() {
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: have we met this problem before?

**Comment by siddontang** [comment]
CI failed 

```
thread 'storage-scheduler::test_high_priority_no_block' panicked at 'write lock is expected for command kv::command::pause keys:(0) 1000 ms | Context { region_id: 0, region_epoch: None, peer: None, term: 0, priority: Normal, isolation_level: SI, not_fill_cache: false, sync_log: false, handle_time: false, scan_detail: false, unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } }', src/storage/txn/scheduler.rs:212:17

note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

stack backtrace:

   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace

             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39

   1: std::sys_common::backtrace::_print

             at src/libstd/sys_common/backtrace.rs:71

   2: std::panicking::default_hook::{{closure}}

             at src/libstd/sys_common/backtrace.rs:59

             at src/libstd/panicking.rs:197

   3: std::panicking::default_hook

             at src/libstd/panicking.rs:211

   4: panic_hook::track_hook::{{closure}}

             at components/panic_hook/src/lib.rs:44

   5: std::thread::local::LocalKey<T>::try_with

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libstd/thread/local.rs:299

   6: std::thread::local::LocalKey<T>::with

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libstd/thread/local.rs:245

   7: panic_hook::track_hook

             at components/panic_hook/src/lib.rs:38

   8: core::ops::function::Fn::call

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libcore/ops/function.rs:69

   9: std::panicking::rust_panic_with_hook

             at src/libstd/panicking.rs:478

  10: std::panicking::continue_panic_fmt

             at src/libstd/panicking.rs:381

  11: std::panicking::begin_panic_fmt

             at src/libstd/panicking.rs:336

  12: tikv::storage::txn::scheduler::Scheduler<E>::enqueue_task

             at src/storage/txn/scheduler.rs:212

  13: tikv::storage::txn::scheduler::Scheduler<E>::schedule_command

             at src/storage/txn/scheduler.rs:267

  14: tikv::storage::txn::scheduler::Scheduler<E>::on_receive_new_cmd

             at src/storage/txn/scheduler.rs:311

  15: <tikv::storage::txn::scheduler::Scheduler<E> as tikv_util::worker::Runnable<tikv::storage::txn::scheduler::Msg>>::run_batch

             at src/storage/txn/scheduler.rs:448

  16: <tikv_util::worker::DefaultRunnerWithTimer<R> as tikv_util::worker::Runnable<T>>::run_batch

             at ./components/tikv_util/src/worker/mod.rs:107

  17: tikv_util::worker::poll

             at ./components/tikv_util/src/worker/mod.rs:267

  18: tikv_util::worker::Worker<T>::start_with_timer::{{closure}}

             at ./components/tikv_util/src/worker/mod.rs:345

test storage::mvcc::reader::scanner::backward::tests::test_many_tombstones ... ok

```

**Comment by siddontang** [comment]
CI failed 

```
thread 'storage-scheduler::test_high_priority_no_block' panicked at 'write lock is expected for command kv::command::pause keys:(0) 1000 ms | Context { region_id: 0, region_epoch: None, peer: None, term: 0, priority: Normal, isolation_level: SI, not_fill_cache: false, sync_log: false, handle_time: false, scan_detail: false, unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } }', src/storage/txn/scheduler.rs:212:17

note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

stack backtrace:

   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace

             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39

   1: std::sys_common::backtrace::_print

             at src/libstd/sys_common/backtrace.rs:71

   2: std::panicking::default_hook::{{closure}}

             at src/libstd/sys_common/backtrace.rs:59

             at src/libstd/panicking.rs:197

   3: std::panicking::default_hook

             at src/libstd/panicking.rs:211

   4: panic_hook::track_hook::{{closure}}

             at components/panic_hook/src/lib.rs:44

   5: std::thread::local::LocalKey<T>::try_with

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libstd/thread/local.rs:299

   6: std::thread::local::LocalKey<T>::with

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libstd/thread/local.rs:245

   7: panic_hook::track_hook

             at components/panic_hook/src/lib.rs:38

   8: core::ops::function::Fn::call

             at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641/src/libcore/ops/function.rs:69

   9: std::panicking::rust_panic_with_hook

             at src/libstd/panicking.rs:478

  10: std::panicking::continue_panic_fmt

             at src/libstd/panicking.rs:381

  11: std::panicking::begin_panic_fmt

             at src/libstd/panicking.rs:336

  12: tikv::storage::txn::scheduler::Scheduler<E>::enqueue_task

             at src/storage/txn/scheduler.rs:212

  13: tikv::storage::txn::scheduler::Scheduler<E>::schedule_command

             at src/storage/txn/scheduler.rs:267

  14: tikv::storage::txn::scheduler::Scheduler<E>::on_receive_new_cmd

             at src/storage/txn/scheduler.rs:311

  15: <tikv::storage::txn::scheduler::Scheduler<E> as tikv_util::worker::Runnable<tikv::storage::txn::scheduler::Msg>>::run_batch

             at src/storage/txn/scheduler.rs:448

  16: <tikv_util::worker::DefaultRunnerWithTimer<R> as tikv_util::worker::Runnable<T>>::run_batch

             at ./components/tikv_util/src/worker/mod.rs:107

  17: tikv_util::worker::poll

             at ./components/tikv_util/src/worker/mod.rs:267

  18: tikv_util::worker::Worker<T>::start_with_timer::{{closure}}

             at ./components/tikv_util/src/worker/mod.rs:345

test storage::mvcc::reader::scanner::backward::tests::test_many_tombstones ... ok

```

File: src/storage/txn/scheduler.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/storage/txn/scheduler.rs]: Yes, i remember the first version of resolve lock forget to acquire latch and caused data corruption.

**Comment by siddontang** [comment]
PTAL @AndreMouche @disksing

**Comment by MyonKeminta** [comment]
Is it covered in tests?

File: src/storage/txn/scheduler.rs (hunk 1 above)
Comment (disksing) [reply src/storage/txn/scheduler.rs]: I think there may be a more elegant way but I cannot come up with an idea. LGTM.

Comment (youjiali1995) [reply src/storage/txn/scheduler.rs]: I also forget to acquire latch for pessimistic locks at the beginning.

**Comment by Connor1996** [comment]
test failed

**Comment by youjiali1995** [comment]
Please fix the conflict.

**Pull request merged**